fem/kernels/kGeometry3D.cpp

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2017, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-734707. All Rights
+// reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+#include "../../general/okina.hpp"
+
+// *****************************************************************************
+MFEM_NAMESPACE
+
+// *****************************************************************************
+template<const int NUM_DOFS,
+         const int NUM_QUAD>
+void kGeom3D(const int numElements,
+             const double* __restrict dofToQuadD,
+             const double* __restrict nodes,
+             double* __restrict J,
+             double* __restrict invJ,
+             double* __restrict detJ)
+{
+   forall(e,numElements,
+   {
+      double s_nodes[3*NUM_DOFS];
```

Comment:
Specific implementation should be used, not sure yet between a global config, or per class switch, or even another idea.

Reply:
As I discussed with you, from a backend developer perspective, I think it's simpler, more flexible, and more extensible to separate the different backend functions as you did rather than in one big monolithic backend where it can be very discouraging to write a backend.